core/src/main/java/org/apache/brooklyn/core/config/ConfigConstraints.java

```diff
@@ -0,0 +1,165 @@
+
+    public static final Logger LOG = LoggerFactory.getLogger(ConfigConstraints.class);
+
+    private final T brooklynObject;
+
+    /**
+     * Checks all constraints of all config keys available to an entity.
+     * <p>
+     * If a constraint is a {@link BrooklynObjectPredicate} then
+     * {@link BrooklynObjectPredicate#apply(Object, BrooklynObject)} will be used.
+     */
+    public static void assertValid(Entity entity) {
+        Iterable<ConfigKey<?>> violations = new EntityConfigConstraints(entity).getViolations();
+        if (!Iterables.isEmpty(violations)) {
+            throw new ConstraintViolationException(errorMessage(entity, violations));
+        }
+    }
+
+    /**
+     * Checks all constraints of all config keys available to an entity adjunct.
+     * <p>
+     * If a constraint is a {@link BrooklynObjectPredicate} then
+     * {@link BrooklynObjectPredicate#apply(Object, BrooklynObject)} will be used.
+     */
+    public static void assertValid(EntityAdjunct adjunct) {
+        Iterable<ConfigKey<?>> violations = new EntityAdjunctConstraints(adjunct).getViolations();
+        if (!Iterables.isEmpty(violations)) {
+            throw new ConstraintViolationException(errorMessage(adjunct, violations));
+        }
+    }
+
+    private static String errorMessage(BrooklynObject object, Iterable<ConfigKey<?>> violations) {
+        StringBuilder message = new StringBuilder("Error configuring ")
+                .append(object.getDisplayName())
+                .append(": [");
+        Iterator<ConfigKey<?>> it = violations.iterator();
+        while (it.hasNext()) {
+            ConfigKey<?> config = it.next();
+            message.append(config.getName())
+                    .append(":")
+                    .append(config.getConstraint());
+            if (it.hasNext()) {
+                message.append(", ");
+            }
+        }
+        return message.append("]").toString();
+    }
+
+    public ConfigConstraints(T brooklynObject) {
+        this.brooklynObject = brooklynObject;
+    }
+
+    abstract Iterable<ConfigKey<?>> getBrooklynObjectTypeConfigKeys();
+
+    public boolean isValid() {
+        return Iterables.isEmpty(getViolations());
+    }
+
+    public Iterable<ConfigKey<?>> getViolations() {
+        return validateAll();
+    }
+
+    @SuppressWarnings("unchecked")
+    private Iterable<ConfigKey<?>> validateAll() {
+        List<ConfigKey<?>> violating = Lists.newArrayList();
+        BrooklynObjectInternal.ConfigurationSupportInternal configInternal = getConfigurationSupportInternal();
+
+        Iterable<ConfigKey<?>> configKeys = getBrooklynObjectTypeConfigKeys();
+        LOG.trace("Checking config keys on {}: {}", getBrooklynObject(), configKeys);
+        for (ConfigKey<?> configKey : configKeys) {
+            // getRaw returns null if explicitly set and absent if config key was unset.
+            Object value = configInternal.getRaw(configKey).or(configKey.getDefaultValue());
```

Comment:
@mikezaccardo what the comment means is that `getRaw()` can return a wrapped null (`Maybe.of(null)`) -- it never returns `null`

but @sjcorbett worth a comment why it's doing `getRaw` + `getDefault` in the first place.  presumably this is to prevent blocking etc.  but maybe that should instead live in a `getNonBlocking` method on `ConfigurationSupportInternal`?  (also it might be okay to allow a small time window in order to evaluate dependent tasks -- cf `EntityConfigResource.get`?)